### PR TITLE
Purge `Oscar.Integer`

### PR DESCRIPTION
**Pull request by thofma**
This is a constant source of confusion and frustration with no benefit. After discussion this with the others, it seems that whatever the intended goal was, it is not something we pursue anymore.

### CODE REVIEW

**Comment by tthsqe12** [comment]
yay! 100% support here

**Review by fingolfin — APPROVED** [review]

File: test/Groups/directproducts.jl [hunk 1]
```diff
@@ -42,7 +42,7 @@
    @test x==H(cperm([1,2]),C[1])
    @test_throws ArgumentError H(cperm([2,3]),C[1])
    @test order(H)==2
-   @test index(G,H) isa Oscar.Integer
+   @test index(G,H) isa Oscar.fmpz
```
Comment (fingolfin) [inline test/Groups/directproducts.jl:45]:
```suggestion
   @test index(G,H) isa fmpz
```

File: src/Groups/matrices/MatGrp.jl [hunk 2]
```diff
@@ -502,9 +502,11 @@ gen(G::MatrixGroup, i::Int) = gens(G)[i]
 
 ngens(G::MatrixGroup) = length(gens(G))
 
-function order(::Type{T}, G::MatrixGroup) where T <: Union{Integer,fmpz}
-   if get_special(G, :order)==nothing return fmpz(BigInt(GAP.Globals.Order(G.X)))
-   else return T(get_special(G, :order))
+function order(::Type{T}, G::MatrixGroup) where T <: Union{Integer, fmpz}
```
Comment (fingolfin) [inline src/Groups/matrices/MatGrp.jl:505]: An alias for `Union{Integer, fmpz}` might be nice, though; perhaps in Nemo?

File: src/Groups/matrices/MatGrp.jl [hunk 3]
```diff
@@ -502,9 +502,11 @@ gen(G::MatrixGroup, i::Int) = gens(G)[i]
...
+   if get_special(G, :order) == nothing
+     return T(BigInt(GAP.Globals.Order(G.X)))::T
```
Comment (fingolfin) [inline src/Groups/matrices/MatGrp.jl:507]: Actually, these days, this should work (but of course you may not want to do that in this PR!) just wanted to mention it
```suggestion
     return T(GAP.Globals.Order(G.X))::T
```

File: src/Groups/matrices/MatGrp.jl (hunk 2 above)
Comment (fingolfin) [reply src/Groups/matrices/MatGrp.jl:505]: In Singular.jl, I added `const IntegerLikeTypes = Union{Integer, n_Z, Nemo.fmpz}` some time ago.

Perhaps `const Intiegerish = Union{Base.Integer, Nemo.fmpz}` or `Intlike` or `IntegerLike` or so...?

I mean, it's not *that* bad to type `Union{Base.Integer,fmpz}` but then I sometimes write `Union{fmpz,Base.Integer}` and it gets a bit tedious to grep for it, and I keep finding places where I should use that Union and so a shortcut would be nice ;-).

**Comment by thofma** [comment]
Regarding the name, should it be also for users or more for internal use? It is a bit annoying that `Vector{IntegerLikeType}` is *not* what one should use. So should it be clear from the type that this is a bit dangerous?

**Comment by fieker** [comment]
On Thu, Sep 09, 2021 at 12:14:20AM -0700, Tommy Hofmann wrote:
> Regarding the name, should it be also for users or more for internal use? It is a bit annoying that `Vector{IntegerLikeType}` is *not* what one should use. So should it be clear from the type that this is a bit dangerous?
> 
There is always the usual problem...

julia> isa([1,2,3], Vector{Integer})
false

julia> isa([1,2,3], Vector{<:Integer})
true

which will not improve with IntegerLikeType - whatever the name

**Comment by thofma** [comment]
Yes, which is why a name like `IntegerUnion` would maybe be more helpful, but it does not look as user friendly. That's why my question.

**Comment by fingolfin** [comment]
I like `IntegerUnion`. For now I envision this for internal use in Nemo, Oscar, Singular, and perhaps some other packages (such as https://github.com/ulthiel/JuLie.jl by @ulthiel). Perhaps at some point users will want to have such a type, too; (whatever "user" exactly means here...) and then we can still add another such type alias if we deem it necessary -- but even then we can still keep `IntegerUnion`, so...

**Pull request merged**